x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/notifications/AbstractAuditMessage.java

```diff
@@ -0,0 +1,195 @@
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalString(resourceId);
+        out.writeOptionalString(message);
+        boolean hasLevel = level != null;
+        out.writeBoolean(hasLevel);
+        if (hasLevel) {
+            level.writeTo(out);
+        }
+        boolean hasTimestamp = timestamp != null;
+        out.writeBoolean(hasTimestamp);
+        if (hasTimestamp) {
+            out.writeLong(timestamp.getTime());
+        }
+        out.writeOptionalString(nodeName);
+    }
+
+    public final String getResourceId() {
+        return resourceId;
+    }
+
+    public final void setResourceId(String resourceId) {
+        this.resourceId = resourceId;
+    }
+
+    public final String getMessage() {
+        return message;
+    }
+
+    public final void setMessage(String message) {
+        this.message = message;
+    }
+
+    public final Level getLevel() {
+        return level;
+    }
+
+    public final void setLevel(Level level) {
+        this.level = level;
+    }
+
+    public final Date getTimestamp() {
+        return timestamp;
+    }
+
+    public final void setTimestamp(Date timestamp) {
+        this.timestamp = timestamp;
+    }
+
+    public final String getNodeName() {
+        return nodeName;
+    }
+
+    public final void setNodeName(String nodeName) {
+        this.nodeName = nodeName;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+        builder.startObject();
+        if (resourceId != null) {
+            builder.field(getResourceField(), resourceId);
+        }
+        if (message != null) {
+            builder.field(MESSAGE.getPreferredName(), message);
+        }
+        if (level != null) {
+            builder.field(LEVEL.getPreferredName(), level);
+        }
+        if (timestamp != null) {
+            builder.field(TIMESTAMP.getPreferredName(), timestamp.getTime());
+        }
+        if (nodeName != null) {
+            builder.field(NODE_NAME.getPreferredName(), nodeName);
+        }
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(resourceId, message, level, timestamp);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (obj == null) {
+            return false;
+        }
+        if (obj instanceof AbstractAuditMessage == false) {
+            return false;
+        }
+
+        AbstractAuditMessage other = (AbstractAuditMessage) obj;
+        return Objects.equals(resourceId, other.resourceId) &&
+            Objects.equals(message, other.message) &&
+            Objects.equals(level, other.level) &&
+            Objects.equals(timestamp, other.timestamp);
+    }
+
+    protected abstract String getResourceField();
+
+    public static final class AuditMessageBuilder<T extends AbstractAuditMessage> {
```

Comment:
Rename to `Builder`?

Reply:
For sure, since it is already nested in that class.